### PR TITLE
Limit reorganization / deep fork depth

### PR DESCRIPTION
**Pull request by gregorycollins**

### CODE REVIEW

File: docs/RecoveringFromDeepForks.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+```
+
+And restart the node with the following changes to `chainweb.yaml`:
+
+```
+chainweb:
+  cuts:
+    pruneChainDatabase: false
+    initialCutHeightLimit: $cut_height_limit
+```
```
Comment (larskuhtz) [inline docs/RecoveringFromDeepForks.md]: one should use the `private` flag and only provide know-peers that are on the "good" target fork.

File: docs/RecoveringFromDeepForks.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+home# systemctl stop chainweb
+home# rm -Rf ~/.local/share/chainweb-node
+home# ssh root@remote-node 'cat chainweb.db.tar.gz' | tar xvzf -
+home# systemctl start chainweb
+```
+
+## Recovery method B: restart node with a cut height limit
+
+This option is slower, but probably safer/easier. The error message will report
+an ancestor blockheight on a particular chain; your goal is to select a
```
Comment (larskuhtz) [inline docs/RecoveringFromDeepForks.md]: Another, more reliable, option is to use the [chainweb-forks](https://github.com/kadena-community/chainweb-forks) tool.

File: docs/RecoveringFromDeepForks.md (hunk 1 above)
Comment (larskuhtz) [reply docs/RecoveringFromDeepForks.md]: This `private` flag is important in deep fork scenarios with "weight inversion". A weight inversion is the situation where a fork exists that has been heavier in the past than the currently heaviest branch.

Normally, this scenario should be even much more rare than deep forks. While deep forks can happen as a consequence of long-lasting network partitions, deep forks with weight inversion can happen as a consequence of long lasting multiple overlapping partitions of the network that got resolved at different times. Forks with weight inversion can also happen on network partitions where the hash power within a partition changes a lot.

In the past forks with weight inversions could also be a consequence of emergency DAs, but that option doesn't exist any more.

File: test/Chainweb/Test/Pact/Utils.hs [hunk 3]
```diff
@@ -397,14 +397,15 @@ testPactCtx v cid bhdb pdb = do
         t0 = BlockCreationTime $ Time (TimeSpan (Micros 0))
     ctx <- TestPactCtx
         <$> newMVar (PactServiceState Nothing mempty 0 t0 Nothing noSPVSupport)
-        <*> pure (PactServiceEnv Nothing cpe pdb bhdb (constGasModel 0) rs True 100000)
+        <*> pure (PactServiceEnv Nothing cpe pdb bhdb (constGasModel 0) rs True 100000 defaultOnFatalError)
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/Utils.hs]: Can this use defaultPactServiceEnv instead

File: test/Chainweb/Test/Pact/PactReplay.hs [hunk 4]
```diff
@@ -78,6 +78,11 @@ tests =
           withTime $ \iot ->
             withPact testVer Quiet pdb bhdb (dupegenMemPoolAccess iot) dir 100000
             (testCase "reject-dupes" . testDupes genblock pdb bhdb)
+        , after AllSucceed "reject-dupes" $
+          withTime $ \iot ->
+            let deepForkLimit = 4 -- lol ... is it possible to get lazier?
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/PactReplay.hs]: remove comment pls

File: src/Chainweb/Pact/PactService.hs [hunk 5]
```diff
@@ -1006,7 +1006,7 @@ execValidateBlock currHeader plData = do
     -- TODO: knob to configure whether this rewind is fatal
     fatalRewindError a h1 h2 = do
         -- TODO: improve error message, give instructions
-        let msg = concat [ "Fatal error: rewind limit exceeded: "
+        let msg = mconcat [ "Fatal error: rewind limit exceeded: "
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: I don't see the point behind this change (and now the indent is wrong)

Comment (emmanueldenloye) [reply src/Chainweb/Pact/PactService.hs]: This comment is about an outdated piece of code.

File: test/Chainweb/Test/Pact/PactReplay.hs (hunk 4 above)
Comment (emmanueldenloye) [reply test/Chainweb/Test/Pact/PactReplay.hs]: ok

File: test/Chainweb/Test/Pact/Utils.hs (hunk 3 above)
Comment (emmanueldenloye) [reply test/Chainweb/Test/Pact/Utils.hs]: ok

File: docs/RecoveringFromDeepForks.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
+# Recovering from deep forks
+
+Help! My node has failed with the following fatal error:
+
+```
+Fatal error: Requested rewind exceeds limit (4). Our previous cut block height: 1, fork ancestor's block height: 6.
+Offending new block:
+BlockHeader {_blockNonce = Nonce 0, _blockCreationTime = BlockCreationTime {_bct = Time (TimeSpan (Micros 1575672858613932))}, _blockParent = "ugAOff6YkKmvp_pJIHqx6HLTRPwyHEHD2c2Olm6L2Fg", _blockAdjacentHashes = BlockHashRecord {_getBlockHashRecord = fromList []}, _blockTarget = HashTarget (PowHashNat 115792089237316195423570985008687907853269984665640564039457584007913129639935), _blockPayloadHash = BlockPayloadHash (MerkleLogHash U-3paUKNKQUkggFmNOnGBEAJnjpy4K_-7t_dGyV0lhw), _blockChainId = ChainId 8, _blockWeight = BlockWeight (HashDifficulty (PowHashNat 1)), _blockHeight = 1, _blockChainwebVersion = fastTimedCPM-peterson, _blockEpochStart = EpochStartTime (Time (TimeSpan (Micros 0))), _blockFlags = FeatureFlags 0, _blockHash = "SKZ0mTWp08ooXsWz_yyT7v_MSrRdyKkmCyWJBTK8u3g"}
+
+Your node is part of a losing fork longer than your reorg-limit, which
```
Comment (larskuhtz) [inline docs/RecoveringFromDeepForks.md]: This statement is not completely correct. For efficiency/DOS reasons we don't validate the forks all the way to the top. So we don't know which fork is actually winning. Doing so would be costly and could be exploited. We thus ask the node administrator to resolve which fork should be given preference.

The [chainweb-forks](https://github.com/kadena-community/chainweb-forks) tool is meant to help administrators to make this decision.

File: src/Chainweb/Pact/Backend/ChainwebPactDb.hs [hunk 7]
```diff
@@ -787,4 +786,4 @@ getEndingTxId bh = callDb "getEndingTxId" $ \db -> do
 -- Careful doing this! It's expensive and for our use case, probably pointless.
 -- We should reserve vacuuming for an offline process
 vacuumDb :: BlockHandler SQLiteEnv ()
-vacuumDb = callDb "vaccumDb" (`exec_` "VACUUM;")
+vacuumDb = callDb "vacuumDb" (`exec_` "VACUUM;")
```
[the remark this answers is not in the document]
Comment (mightybyte) [reply src/Chainweb/Pact/Backend/ChainwebPactDb.hs]: This change has no impact.  It only changes a log message.

**Pull request merged**